bolt-modules/boltlib/lib/puppet/functions/wait.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+require 'bolt/logger'
+require 'bolt/target'
+
+# Wait for a Future or array of Futures to finish and return results,
+# optionally with a timeout.
+#
+# > **Note:** Not available in apply block
+Puppet::Functions.create_function(:wait, Puppet::Functions::InternalFunction) do
+  # Wait for a Future or array of Futures to finish.
+  # @param futures A Bolt Future object or array of Bolt Futures to wait on.
+  # @option options [Boolean] _catch_errors Whether to catch raised errors.
+  # @return A Result or Results from the Futures
+  # @example Upload a large file in the background, then wait until it's loaded
+  #   $futures = background() || {
+  #     upload_file("./very_large_file", "/opt/jfrog/artifactory/var/etc/artifactory", $targets)
+  #   }
+  #   # Run an unrelated task
+  #   run_task("deploy", $targets)
+  #   # Wait for the file upload to finish
+  #   $results = wait($futures)
+  dispatch :wait do
+    param 'Variant[Future, Array[Future]]', :futures
+    optional_param 'Hash[String[1], Any]', :options
+    return_type 'Array[Boltlib::PlanResult]'
+  end
+
+  # Wait for a Future or array of Futures to finish, timing out on any Futures
+  # don't finish within the provided timeout. Returns the results of the Futures.
```

Comment:
This feels a bit verbose for a section header. Here's how it's rendered in docs:

![image](https://user-images.githubusercontent.com/30373575/119199417-dc7fc880-ba3f-11eb-84b2-0f80cdca2b64.png)

I would have the description at the top of the file be the verbose description and the dispatch descriptions be simple, one-line summaries (e.g. "Wait for Futures to finish with a timeout").